cli/src/diff_util.rs

```diff
@@ -1050,40 +1050,39 @@ pub fn show_diff_stat(
         max_bar_length as f64 / max_diffs as f64
     };
 
-    formatter.with_label("diff", |formatter| {
-        let mut total_added = 0;
-        let mut total_removed = 0;
-        let total_files = stats.len();
-        for stat in &stats {
-            total_added += stat.added;
-            total_removed += stat.removed;
-            let bar_added = (stat.added as f64 * factor).ceil() as usize;
-            let bar_removed = (stat.removed as f64 * factor).ceil() as usize;
-            // replace start of path with ellipsis if the path is too long
-            let (path, path_width) = text_util::elide_start(&stat.path, "...", max_path_width);
-            let path_pad_width = max_path_width - path_width;
-            write!(
-                formatter,
-                "{path}{:path_pad_width$} | {:>number_padding$}{}",
-                "", // pad to max_path_width
-                stat.added + stat.removed,
-                if bar_added + bar_removed > 0 { " " } else { "" },
-            )?;
-            write!(formatter.labeled("added"), "{}", "+".repeat(bar_added))?;
-            writeln!(formatter.labeled("removed"), "{}", "-".repeat(bar_removed))?;
-        }
-        writeln!(
-            formatter.labeled("stat-summary"),
-            "{} file{} changed, {} insertion{}(+), {} deletion{}(-)",
-            total_files,
-            if total_files == 1 { "" } else { "s" },
-            total_added,
-            if total_added == 1 { "" } else { "s" },
-            total_removed,
-            if total_removed == 1 { "" } else { "s" },
+    formatter.push_label("diff")?;
```

Comment:
I wonder, does `formatter.with_label` have to take a closure? I wonder if we could often use a block instead of a closure:

```rust
let formatter=...;

{
  let formatter = formatter.with_label("diff"); // Or let diff_formatter = ...
  ...
}
```

There could also be a `with_label_do` function (perhaps with a better name) that works like `with_label` does now.

Reply:
I've attempted that. The problem is that we can't safely pop the label in the `Drop` implementation because popping the label can result in new output, which can fail, and `Drop::drop()` has no way to propagate the error.

Reply:
`.with_label()` can be changed to accept generic error type, but that can be addressed separately.

Reply:
Interesting, this explains a lot! I'm glad Yuya sees a workaround for the problem that actually matters.

Now definitely outside the scope of this PR, just curious:

- We couldn't just ignore errors inside `Drop::drop`, could we?
- This is a confusing problem, it seems really painful for many types of RAII. I wonder how crates like `tempfile` work around it. I guess they just ignore errors; I'll probably look at them unless you just know how they work and tell me.

Reply:
I actually started working on a PR (again) for this. I'll share it shortly and you can decide what you think. It ignores errors in `Drop::drop`, yes.